### ISSUE TITLE
Processing of reconciliation is slow?

### ISSUE THREAD

**Issue by gitonthescene** [OpenRefine/OpenRefine#3629]
As discussed on [Gitter](https://gitter.im/OpenRefine/OpenRefine), I'm developing a [reconciliation service](https://github.com/gitonthescene/csv-reconcile) and I noticed that the time it takes to do the fuzzy search is dominated by the time it takes to process the results.

I'm not sure if you have access to Python, but I've tried to make the [instructions](https://github.com/gitonthescene/csv-reconcile) as straightforward as possible.  It should work on any platform as long as you set the environment properly.  Also, it's currently set up so you can lower the log level just by pointing `FLASK_APP` at `venv/lib/python3.7/site-packages/csv_reconcile` instead of `csv_reconcile`.

I just threw this up so you could test it.  The tool could use some more polishing.

You can load `sample/progressives.tsv` into OpenRefine and then reconcile the `member` column using the service.  On my machine OpenRefine 3.4.1 sends query batches of size 10.  The service takes <2 seconds to process, but there are ~10 secs between batches getting sent.

``` sh
bash-5.0$ venv/bin/python -m flask run
 * Serving Flask app "venv/lib/python3.7/site-packages/csv_reconcile"
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
start timer
Elapsed: 1.5238202030000014
127.0.0.1 - - [13/Feb/2021 19:45:17] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.5245007290000103
127.0.0.1 - - [13/Feb/2021 19:52:42] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.5689725690000387
127.0.0.1 - - [13/Feb/2021 19:52:44] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.5700575550000053
127.0.0.1 - - [13/Feb/2021 19:52:52] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.7569187119999583
127.0.0.1 - - [13/Feb/2021 19:52:57] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.4609496090000107
127.0.0.1 - - [13/Feb/2021 19:53:09] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.8019815920000042
127.0.0.1 - - [13/Feb/2021 19:53:15] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.667001018999997
127.0.0.1 - - [13/Feb/2021 19:53:18] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.6668618590000506
127.0.0.1 - - [13/Feb/2021 19:53:29] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.7560411690000137
127.0.0.1 - - [13/Feb/2021 19:53:36] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 1.6570629999999937
127.0.0.1 - - [13/Feb/2021 19:53:39] "POST /reconcile HTTP/1.1" 200 -
start timer
Elapsed: 0.5511451229999693
127.0.0.1 - - [13/Feb/2021 19:53:50] "POST /reconcile HTTP/1.1" 200 -
```

Hmm...  A few of the batches above are faster than I had been seeing.  Not sure why it wouldn't be uniform.  In any event, this is how I produced the results.  Maybe it's affected by other projects running in OpenRefine??  I'm happy to provide more information if you need it.

Regards,

**Comment by gitonthescene** [comment]
Hmm..  Strangely it looks like I need to use JSONP for this service only to add the service.  Once the service is registered that seems not to be required.  Note that this is a service running on local host so that [CORS shouldn't be required](https://github.com/reconciliation-api/specs/issues/64).  I can set up CORS, but that means setting up an ssh certificate, etc., which is a bit of a pain as a pre-requisite to development.

If I don't use JSONP, I get this when adding the service, but *only* when adding the service.

<img width="874" alt="Screen Shot 2021-02-15 at 1 58 47 PM" src="https://user-images.githubusercontent.com/45531702/107907123-287d2600-6f96-11eb-8477-10b41cc67c8f.png">

Please let me know if you've had issues with this.  I'll update the code to use JSONP for now.

Regards,

**Comment by gitonthescene** [comment]
Actually, let alone development, simply running a local service shouldn't require setting up local ssh certificates, etc.  This app for instance is meant to be run locally to be able to reconcile against a CSV file.

**Comment by gitonthescene** [comment]
Sorry.  I still think it strange that JSONP is required only for adding the service when not using CORS, but I was wrong about CORS requiring ssh.  With [flask-cors](https://flask-cors.readthedocs.io/en/3.0.10/) it was easy enough to implement, so I've removed JSONP and added CORS.

**Comment by tfmorris** [comment]
Yes, I got blocked on the JSONP issue when I tried to profile the code and gave up. I'll try to circle back some time to take another look.

**Comment by gitonthescene** [comment]
I'm on a mac and don't know a better way to monitor activity than using top.  It looks like you can see where the Python service is doing the fuzzy calculation, but other than that it looks like not a lot is going on.  Ideally I'd like to get my results back sooner.  Any help understanding what's taking so long to get subsequent requests would be appreciated.

FWIW, if this is trying to be nice to the server, I'd think it better for the server to throttle and the client to handle getting throttled with a back off rather than slowing down requests for every client usage regardless of how well the service is performing.

```
$ egrep 'Python|JavaAppLauncher' /tmp/monitor.script
74728  Python          0.0  1   sleeping 00:06.18
26164  JavaAppLauncher 0.0  53  sleeping 02:55.73
74728  Python          13.4 2/1 running  00:06.51
26164  JavaAppLauncher 0.2  54  sleeping 02:55.73
74728  Python          14.6 1   sleeping 00:06.70
26164  JavaAppLauncher 0.0  54  sleeping 02:55.73
26164  JavaAppLauncher 0.0  54  sleeping 02:55.73
74728  Python          0.0  1   sleeping 00:06.70
26164  JavaAppLauncher 0.0  54  sleeping 02:55.73
74728  Python          0.0  1   sleeping 00:06.70
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.70
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.70
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.70
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.70
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.70
74728  Python          19.0 1   sleeping 00:06.93
26164  JavaAppLauncher 0.1  54  sleeping 02:55.74
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.93
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.93
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.93
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.93
26164  JavaAppLauncher 0.0  54  sleeping 02:55.74
74728  Python          0.0  1   sleeping 00:06.93
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:06.93
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:06.93
74728  Python          0.1  2/1 running  00:06.94
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          19.2 1   sleeping 00:07.17
26164  JavaAppLauncher 0.1  54  sleeping 02:55.75
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:07.17
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:07.17
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:07.17
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:07.17
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:07.17
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:07.17
26164  JavaAppLauncher 0.0  54  sleeping 02:55.75
74728  Python          0.0  1   sleeping 00:07.17
74728  Python          19.1 1   sleeping 00:07.40
26164  JavaAppLauncher 0.1  54  sleeping 02:55.76
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
26164  JavaAppLauncher 0.0  54  sleeping 02:55.76
74728  Python          0.0  1   sleeping 00:07.40
74728  Python          19.2 1   sleeping 00:07.63
26164  JavaAppLauncher 0.1  54  sleeping 02:55.76
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.63
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.63
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.63
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.63
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.63
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.63
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.63
74728  Python          18.5 1   sleeping 00:07.87
26164  JavaAppLauncher 0.1  54  sleeping 02:55.77
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.87
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.87
26164  JavaAppLauncher 0.0  54  sleeping 02:55.77
74728  Python          0.0  1   sleeping 00:07.87
26164  JavaAppLauncher 0.0  54  sleeping 02:55.78
74728  Python          0.0  1   sleeping 00:07.87
26164  JavaAppLauncher 0.0  54  sleeping 02:55.78
74728  Python          0.0  1   sleeping 00:07.87
$
```

**Comment by tfmorris** [comment]
I'm able to reproduce this with 3.4.1, but the HTTP code has been completely rewritten for 3.5, so I'm not sure it's worth spending a lot of time debugging the problem. The problem doesn't exist in 3.5, although there's another problem (https://github.com/OpenRefine/OpenRefine/issues/3664) with reusing HTTP connections that causes every other reconciliation batch to fail.

I profiled this and the time spent calculating features for your example is in the range 1-5 msec for a batch of ten rows, so it's definitely not a bottleneck.

**Comment by gitonthescene** [comment]
That’s fair enough.  That range sounds much more workable.  If I see the problem in 3.5 I’ll open a new issue.  Feel free to close this one.

**Comment by wetneb** [comment]
Closing then, but I keep in mind your suggestion of lazy computation for recon features, which makes a lot of sense I think (not just for time savings, but also for memory and disk space).

**Comment by gitonthescene** [comment]
FWIW, the other idea I had is to move Levenshtein distance from being a feature to being a “data extension”.  With that viewpoint, you could imagine it being a “plugin”, meaning a data extension without requiring a service.  That notion could be extended to allow people to write their own plugins for functionality which suits their domain.  Also as a data extension it’s more explicitly “on demand”.

**Comment by gitonthescene** [comment]
Just had another simpler idea.  Both Levenshtein distance and Word distance can be exposed as functions so that users can call them as they see fit.  For instance, you could split out all the candidates returned from reconciliation and then apply Levenshtein distance with a GREL expression.  Are these functions already available?  The links to GREL controls, Clojure and Jython seem broken [here](https://docs.openrefine.org/manual/expressions/#grel-controls).

**Comment by gitonthescene** [comment]
I'm no Clojure wiz, but I got the following to work.  It might be nice to expose it as a function, though.  Also, it might be nice if the documentation had some more examples of Clojure.  I found [this](https://kirank.in/posts-output/2016-08-20-openrefine-clojure/) useful.

``` clojure
(defn getCol [colnm]
  (.-value (.-cell (.getField cells colnm (new java.util.Properties)))))
(import [org.apache.commons.text.similarity LevenshteinDistance])
(.apply (new LevenshteinDistance) (getCol "col") (getCol "coord"))
```

**Comment by wetneb** [comment]
Makes sense! If you feel like it, I am sure PRs to the docs in this direction will be welcome :)

**Comment by gitonthescene** [comment]
For posterity, this is even easier with Jython.  (I have no prior experience with either...)

```from org.apache.commons.text.similarity import LevenshteinDistance
ld = LevenshteinDistance()
return ld.apply(cells.col.value, cells.coord.value)
```

**Comment by gitonthescene** [comment]
@wetneb Added doc about basic Clojure access [here](https://github.com/OpenRefine/OpenRefine/wiki/Recipes#11-clojure).

**Comment by gitonthescene** [comment]
I don't know if this should be a new issue or not, but reconciliation feels slow again.  Above @tfmorris says that this went away in OpenRefine 3.5 and I seem to remember confirming at the time, but that was before the release.  I've now tested with both OpenRefine 3.5 and 3.6.1 stable releases and both feel slow.  I can see the time spent in the server and its negligible so it looks like the client either isn't processing results quickly or is throttling requests for some reason.

I thought @antoine2711 made some change to batch requests, but I can't seem to find the issue now.

**Comment by gitonthescene** [comment]
P.S. I'm happy to open a new issue if that's preferred.  I just wanted to draw the connection to the old issue.
